plugin/completion.py

```diff
@@ -72,274 +111,81 @@ def initialize(self) -> None:
                             completion_triggers.append({
-                                'characters': "".join(self.trigger_chars),
+                                'characters': "".join(trigger_chars),
                                 'selector': scope
                             })
 
             self.view.settings().set('auto_complete_triggers', completion_triggers)
 
-    def is_after_trigger_character(self, location: int) -> bool:
-        if location > 0:
-            prev_char = self.view.substr(location - 1)
-            return prev_char in self.trigger_chars
-        else:
-            return False
+    def on_query_completions(self, prefix: str, locations: List[int]) -> Optional[sublime.CompletionList]:
+        if not self.initialized:
+            self.initialize()
 
-    def is_same_completion(self, prefix: str, locations: List[int]) -> bool:
-        if self.response_incomplete:
-            return False
+        if not self.enabled:
+            return None
 
-        if self.last_location < 0:
-            return False
+        completion_list = sublime.CompletionList()
 
-        # completion requests from the same location with the same prefix are cached.
-        current_start = locations[0] - len(prefix)
-        last_start = self.last_location - len(self.last_prefix)
-        return prefix.startswith(self.last_prefix) and current_start == last_start
-
-    def find_completion_item(self, inserted: str) -> Optional[dict]:
-        """
-
-        Returns the completionItem for a given replacement string.
-        Matches exactly or up to first snippet placeholder ($s)
-
-        """
-        # TODO: candidate for extracting and thorough testing.
-        if self.completions:
-            for index, item in enumerate(self.completions):
-                trigger, replacement = item
-
-                snippet_offset = replacement.find('$', 2)
-                if snippet_offset > -1:
-                    if inserted.startswith(replacement[:snippet_offset]):
-                        return self.response_items[index]
-                else:
-                    if replacement == inserted:
-                        return self.response_items[index]
-        return None
-
-    def on_modified(self) -> None:
-
-        # hide completion when backspacing past last completion.
-        if self.view.sel()[0].begin() < self.last_location:
-            self.last_location = -1
-            self.view.run_command("hide_auto_complete")
-
-        # cancel current completion if the previous input is an space
-        prev_char = self.view.substr(self.view.sel()[0].begin() - 1)
-        if self.state == CompletionState.REQUESTING and prev_char.isspace():
-            self.state = CompletionState.CANCELLING
-
-        if self.committing:
-            self.committing = False
-            self.on_completion_inserted()
+        # this is for tests
+        if self.test_completions:
```

Comment:
I really would not like to add code like this just to make this testable. 
Can we take a look on why the self.set_response and self.await_message('textDocument/completion') don't work as expected in tests? 🙂

Reply:
I took the time to check out the changes locally.

This works for me:

```python
    def test_simple_label(self) -> 'Generator':
        self.set_response("textDocument/completion", label_completions)
        self.type("a")
        count = self.view.change_count()
        yield from self.select_completion()
        yield from self.await_message("textDocument/completion")
        yield from self.await_view_change(count + 2)
        self.assertEquals(self.read_file(), 'asdf')
```

Reply:
Thanks for the help. :)
I updated all the tests. I didn't use `await_view_change` because I think that the tests look simpler without it.

BTW, I am so happy to see the completions tests working again :)
I only need to update the test_completion_core.py and to be done with the tests.(and after that I can consider this PR as done, unless you have further change requests)